community/kernel/src/main/java/org/neo4j/kernel/api/cursor/NodeItem.java

```diff
@@ -0,0 +1,166 @@
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.api.cursor;
+
+import java.util.Arrays;
+
+import org.neo4j.collection.primitive.PrimitiveIntIterator;
+import org.neo4j.cursor.Cursor;
+import org.neo4j.function.IntSupplier;
+import org.neo4j.function.ToIntFunction;
+import org.neo4j.graphdb.Direction;
+import org.neo4j.kernel.impl.api.store.CursorRelationshipIterator;
+import org.neo4j.kernel.impl.api.store.RelationshipIterator;
+import org.neo4j.kernel.impl.util.Cursors;
+
+/**
+ * Represents a single node from a cursor.
+ */
+public interface NodeItem
+        extends EntityItem
+{
+    public abstract class NodeItemHelper
+            extends EntityItemHelper
+            implements NodeItem
+    {
+        ToIntFunction<LabelItem> GET_LABEL = new ToIntFunction<LabelItem>()
+        {
+            @Override
+            public int apply( LabelItem item )
+            {
+                return item.getAsInt();
+            }
+        };
+
+        ToIntFunction<IntSupplier> GET_RELATIONSHIP_TYPE = new ToIntFunction<IntSupplier>()
+        {
+            @Override
+            public int apply( IntSupplier item )
+            {
+                return item.getAsInt();
+            }
+        };
+
+        @Override
+        public boolean hasLabel( int labelId )
+        {
+            try ( Cursor<LabelItem> labelCursor = label( labelId ) )
+            {
+                return labelCursor.next();
+            }
+        }
+
+        @Override
+        public PrimitiveIntIterator getLabels()
+        {
+            return Cursors.intIterator( labels(), GET_LABEL );
+        }
+
+        @Override
+        public RelationshipIterator getRelationships( Direction direction, int[] relTypes )
+        {
+            relTypes = deduplicate( relTypes );
+
+            return new CursorRelationshipIterator( relationships( direction, relTypes ) );
+        }
+
+        public RelationshipIterator getRelationships( Direction direction )
+        {
+            return new CursorRelationshipIterator( relationships( direction ) );
+        }
+
+        @Override
+        public PrimitiveIntIterator getRelationshipTypes()
+        {
+            return Cursors.intIterator( relationshipTypes(), GET_RELATIONSHIP_TYPE );
+        }
+
+        private static int[] deduplicate( int[] types )
```

Comment:
I agree, for example cypher plans can deduplicate once in the planning and then swear to always pass in proper type arrays to the kernel api.